### PR TITLE
Use read-buffering on `urandom`

### PR DESCRIPTION
**Pull request by jgaskins**
I assume that disabling read buffering on `/dev/urandom` was intended to avoid hogging random entropy. I’m calling it an assumption because I can’t find mention of it in https://github.com/crystal-lang/crystal/pull/2959 where it was added, but playing nice with the system entropy device is a good idea, so it probably seemed reasonable. However, the computer can spare 8KB of entropy per process (it’s at most 8KB wasted per process) and anything that uses `Random::Secure` heavily ends up bottlenecked on `File#read`. And it’s a _big_ bottleneck, too — approximately 1836% slower:

Before:

    UUID.random   1.06M (939.51ns) (± 0.54%)  0.0B/op

After:

    UUID.random  19.46M ( 51.38ns) (± 1.64%)  0.0B/op

I discovered this while writing my own entropy buffer (taking 1µs for every `Random::Secure` invocation seemed like a lot longer than it should take), which defeats the purpose of not buffering. A single process-wide entropy buffer is preferable to multiple consumers of `Random:::Secure` implementing separate buffers.

---

Could we do the same for Linux, where most of us are deploying our apps anyway? I’m curious what the performance impact would be of buffering `getrandom` calls for Linux. I imagine it would be quite similar to this since it also involves jumping between user and kernel space just like `File#read` does. I’d have to try it out on a cloud VM maybe next week or so, but the buffer would look something like this (writing it in here so I don’t forget):

```crystal
module Crystal::System::Random
  GENERATOR = SecureGenerator.new

  private class SecureGenerator
    getter buffer : Bytes = Bytes.new(8192).tap { |b| sys_getrandom b }
    getter index = 0

    def bytes(bytes : Bytes) : Nil
      if (offset = bytes.bytesize - (buffer.bytesize - index)) > 0
        (buffer + index).copy_to bytes
        @index = 0
        sys_getrandom buffer
        bytes(bytes + offset)
      else
        (buffer + index).copy_to bytes.to_unsafe, bytes.bytesize
        @index += bytes.bytesize
      end
    end
  end
end
```

`sys_getrandom` is [this method](https://github.com/crystal-lang/crystal/blob/f6f92a15a8df4637ebbc29c807d56ccb995fd1a4/src/crystal/system/unix/getrandom.cr#L101-L115).

### CODE REVIEW

**Comment by jhass** [comment]
How do comparable abstractions in other languages handle this? 

I'm a little worried storing randomness for an extended time rather than getting fresh one when actually needed could slightly increase predictably in some scenarios. But that might be overly cautious.

**Comment by straight-shoota** [comment]
I share these concerns. It doesn't seem like a good idea to cache random data in process memory. It pretty certainly deteriorates the security value of `Random::Secure`.

So I wouldn't touch that.

However, it should be easy to improve the performance characteristics of random data generation by user choice. `UUID.random` receives a `Random` source as parameter. That's the strongest one available by default (`Random::Secure`). But you can easily exchange it for more performant implementations, if the trade-off in security is negligible for the specific use case. That doesn't need to go down to a PRNG, but it should be quite trivial to implement a buffered random source on top of `Random::Secure`, for example.

**Comment by lbguilherme** [comment]
One critical issue of buffering is that after a `fork` both processes will generate the same random data for a while. At the very least, there should be either a `pthread_atfork` callback or a explicit check for the current PID on every call.

Another concern is that an arbitrary memory read vulnerability anywhere in the process can lead to the random buffer leaking and a privilege escalation out of it, no good for something that has "secure" in the name.

- BoringSSL has a off-by-default option to enable buffering. As far as I can tell other SSL options don't have such option (none of them buffer by default)
- The Java VM has buffering, but it is only 32 bytes in size and only buffer for 100ms, after that the buffer is discarded.

This isn't a simple change.

**Comment by jgaskins** [comment]
@straight-shoota Can you clarify how Crystal supports forking if the public API for it was removed? On the surface, those two things seem to be in conflict.

**Comment by kostya** [comment]
if you need fast UUID.random, why not call it with another random generator? as first argument?

**Comment by straight-shoota** [comment]
You can always call `LibC.fork`.

**Comment by straight-shoota** [comment]
@kostya I suppose the question is what should be the default behaviour. The default should definitely prioritize security over speed. But speed is great, of course. So it could be possible to accept negligible compromises in security for improved performance.

**Comment by lbguilherme** [comment]
Go does not support fork at all, so it was easier to implement. But the fact that they do buffer the random data is very interesting, maybe the security implication isn't that big after all, I'm not sure.

About forking, it is possible to detect when a fork happens using `pthread_atfork` on every thread. As we support forking, we probably need to implement this. Or maybe as the user is forking with a very low-level function from LibC, perhaps it's ok to assume the user knows what she is doing?

I have more questions than answers right now.

**Comment by jgaskins** [comment]
> if you need fast UUID.random, why not call it with another random generator? as first argument?

@kostya This would require that I implement my own `Random` interface to the CSPRNG provided by the OS. `UUID` is only a use case I used for the benchmark, but it’s not _the_ thing I’m talking about.

> You can always call `LibC.fork`.

@straight-shoota That’s not the same as supporting forking, though. If I were to say that [`LibC.pthread_create`](https://github.com/crystal-lang/crystal/blob/ef05e26d6ecb0c7d1f5f0568af76ed001896a601/src/lib_c/x86_64-darwin/c/pthread.cr#L14) causes problems in my app, would you still say “Crystal supports that”?

> Or maybe as the user is forking with a very low-level function from LibC, perhaps it's ok to assume the user knows what she is doing?

@lbguilherme I agree with this, especially considering `fork` was deliberately removed. That’s a very different concept from it simply not yet being added.

**Comment by asterite** [comment]
You can also fork in Go if you bind to the C fork function.

With that said: no, Crystal does not support fork. Yes, you can do it. But we shouldn't design the language and API around something we don't want users to do.

**Comment by straight-shoota** [comment]
FTR: My comments on `fork` were just to clarify the situation. I don't think it's a blocker for buffering random data. It's definitely possible to work around potential issues with that.

I suppose my main concern for buffering is that random data will be stored in memory and thus may be accessible there for a long amount of time, which creates opportunities for adversaries to read it. This could even be used to "predict" randomized data.

**Comment by didactic-drunk** [comment]
Will this PR cause a single `Random` call to read 8k from the system random pool?

If so then crystal unix like utilities that quickly exit after performing a single task *may* discard most of the read buffer and/or starve the system of randomness. That may be a poor default.

Most programs don't need large amounts of high quality randomness.  Rarer still is having the RNG as the bottleneck.  Maybe your program needs a different solution?

I also have need large amounts of random data (gigabytes per run). [My solution](https://didactic-drunk.github.io/sodium.cr/Sodium/Cipher/Chalsa.html#random_bytes(bytes:Bytes):Bytes-instance-method) involves using a frequently reseeded [stream cipher as the RNG](https://en.wikipedia.org/wiki/Cryptographically-secure_pseudorandom_number_generator#Designs_based_on_cryptographic_primitives).

**Comment by jgaskins** [comment]
@didactic-drunk I discovered that randomness doesn’t work like that anymore when I was looking into [this](https://github.com/crystal-lang/crystal/pull/11974#issuecomment-1090315259) the other day. A lot of our old practices of being nice to the entropy pool have apparently been overkill for the past few years — since Linux kernel 5.6 and macOS Catalina, real entropy is only used to seed a PRNG. In concrete terms: `/dev/random` and `/dev/urandom` now have identical functionality in both OSes.

**Comment by lbguilherme** [comment]
This is a great read on the topic: https://www.2uo.de/myths-about-urandom/

**Comment by rdp** [comment]
What benchmark did you use for the comparison, what OS?  I remember being surprised by how slow Random was on OS X once...UUID's were the bottleneck in a web server benchmark IIRC once...there should at least be an option to use cached there, I imagine...

**Comment by dmgk** [comment]
Just as a side note, all buffering [was removed](https://github.com/golang/go/commit/3ae414c31e59c9ee210fa3606f36cf0ea36b9906) from Go crypto/rand.

**Comment by jgaskins** [comment]
> Just as a side note, all buffering [was removed](https://github.com/golang/go/commit/3ae414c31e59c9ee210fa3606f36cf0ea36b9906) from Go crypto/rand.

@dmgk Interesting, that was only a week or two after I opened this PR. Notably, though, it looks like they did it only because their buffering involves mutex locks, but to my knowledge the Crystal implementation does not.

Do we know if Crystal would need to use mutex locks to support the multithreaded fiber scheduler?

Also, do we know if Go uses POSIX mutexes or userspace ones? Crystal’s `Mutex` class locks and unlocks 2 orders of magnitude faster than generating a UUID on my machine (10-12ns vs 1µs) because it’s all done in userspace, whereas generating a UUID requires dipping into kernelspace, so even if we added a mutex lock this PR should still be significantly faster than the current implementation, though that would need to be checked.

> What benchmark did you use for the comparison, what OS? I remember being surprised by how slow Random was on OS X once...UUID's were the bottleneck in a web server benchmark IIRC once...there should at least be an option to use cached there, I imagine...

@rdp I was just running `UUID.random` in a `Benchmark.ips` loop on macOS. I’m surprised I didn’t include the benchmark code — I always try to. The code would’ve looked something like this:

```crystal
require “benchmark”
require “uuid”

Benchark.ips do |x|
  x.report { UUID.random }
end
```

I never did end up running it on Linux, though.

**Comment by dmgk** [comment]
@jgaskins AFAIK, all Go synchronization primitives are implemented with atomics (from runtime∕internal∕atomic package), there's no kernel involved. But I think Go is perhaps not the right role model for this particular PR because Go doesn't generally use interfaces provided by the libc (unless it's unavoidable as with OpenBSD where all syscalls have to go through libc). Crystal doesn't have these restrictions and it's sometimes possible to switch to better interfaces.

As an example, there's https://github.com/crystal-lang/crystal/pull/12608 that switches `Crystal::System::Random` from `/dev/urandom` to `arc4random(3)`. Running the benchmark above, I'm getting these results on my FreeBSD box:

with `/dev/urandom`:
```
$ ./uuid
   1.35M (740.57ns) (± 0.45%)  0.0B/op  fastest
```

with `arc4random(3)`:
```
$ ./uuid
  18.89M ( 52.93ns) (± 0.46%)  0.0B/op  fastest
```

I'm not very familiar with modern Darwin, perhaps they also provide `arc4random(3)` or similar in their libc and we can just switch to it?

**Comment by rdp** [comment]
On my linux crystal 1.2 on WSL:

Normal:

 252.73k (  3.96µs) (± 1.48%)  0.0B/op  fastest

with caching snippet provided (no mutexes):

 253.65k (  3.94µs) (± 1.23%)  0.0B/op  fastest

with caching snippet provided (mutexes approximated like `  x.report { m.synchronize {UUID.random }}`, I assume mutexes would be needed eventually)

 245.10k (  4.08µs) (± 1.42%)  0.0B/op  fastest

Interested in a speedup for OS X, either this or arc4random (unfortunately don't have an OS X box to test on at present...).

**Comment by jgaskins** [comment]
@dmgk Excellent find! I didn't realize `arc4random` was *that* much faster. It does work on Darwin, and on an Intel Mac it's very nearly as fast as this PR:

Crystal 1.6.1 release:

```crystal
uuid   1.10M (908.59ns) (± 1.66%)  0.0B/op  fastest
```

Using `arc4random` (implementation taken from #12608:

```
uuid  15.80M ( 63.27ns) (± 1.49%)  0.0B/op  fastest
```

<details><summary>Benchmark code</summary>

```crystal
require "uuid"
require "benchmark"

# Smoke test UUID generation
10.times { pp UUID.random }

Benchmark.ips do |x|
  x.report { UUID.random }
end

lib LibC
  fun arc4random : UInt32
  fun arc4random_buf(x0 : Void*, x1 : SizeT) : Void
  fun arc4random_uniform(x0 : UInt32T) : UInt32T
end

module Crystal::System::Random
  # Fills *buffer* with random bytes using arc4random.
  def self.random_bytes(buffer : Bytes) : Nil
    LibC.arc4random_buf(buffer.to_unsafe.as(Void*), buffer.size)
  end

  def self.next_u : UInt32
    LibC.arc4random
  end
end
```

</details>

The speedup on my ARM Mac is only 3.5x with the same patch and is actually slower than the Intel version, so I'm not sure if that indicates a bottleneck within Crystal on `aarch64-darwin` or if `arc4random_buf` just isn't as fast on ARM. However, it was significantly faster than the Intel version *without* the patch (2.5-3x, I believe), and it's still an improvement.

TL;DR: If we add `arc4random` support to #12608 for Darwin this PR becomes obsolete, so I'll close this.
